### ISSUE TITLE
Limit number of concurrent jobs

### ISSUE THREAD

**Issue by dadarakt** [ChristopherHX/github-act-runner#66]
### TLDR; 
Would it be possible to limit the amount of concurrent jobs a runner takes on?

## Our setup
We employ a single runner for multiple repositories within our organisation.
Both are Elixir projects and we noticed that the runner will happily run multiple of those jobs in parallel. 
This will cause a problem in fetching the dependencies, as one process is already locking the tool (mix), thus failing any concurrent runs except the first one.

We can limit concurrency in the workflow for one repository, but this doesn't seem to be possible across multiple repositories.

I think that such problems could occur in any setup where some resources are limited to one process accessing them, so maybe it's interesting for others as well.

Happy for any input, maybe there's a workaround I'm not aware of?

**Comment by ChristopherHX** [comment]
Concurrency should be limited to 1 for multi repository runners, you found a race condition in my code.
This line is expected to stop all Message listeners, but it seems like this section needs a concurrency lock.

https://github.com/ChristopherHX/github-act-runner/blob/516c04ae84cd895345ad995f91a6ffcbd8fc6078/main.go#L931

_Hmm the behavior of the github actions service changed, ignoring the job requests for some minutes causes a job failure. Turning down communtication to github seems like a bad idea as of now._

**Comment by dadarakt** [comment]
Thanks for picking this up!
Yes indeed, causing a failure by ignoring the communication doesn't seem like a great idea either. Is there a way to acknowledge the job request, but refusing to accept it?

**Comment by ChristopherHX** [comment]
> Is there a way to acknowledge the job request, but refusing to accept it?

I honestly don't know. I wrote this runner just for fun and have no idea what exactly github does on their server. At least they didn't make a breaking change to the runner protocol since a year, so communication can be considered fairly stable.

It is possible to accept a job, but just send keep alive logs / api and wait for the previous job to finish. This still can be a problem if you have set a job timeout, because this kind of waiting counts toward job execution time.

We would need to test this over weeks / months, then we know if we have to keep polling all repos for jobs to avoid job failures.

You can try https://github.com/ChristopherHX/github-act-runner/pull/67, I was experimenting with external worker processes for job execution adding a mutex lock wasn't much efford. You find prebuilds under the github actions checks of the PR.
